### DISCUSSION TITLE
Script URLs with authentication (BasicAuth, OAuth2)

### DISCUSSION THREAD

**Discussion by arittner** [plantuml/plantuml#783] [Ideas]
Hello!

_(I tried to write a forum message, but I gave up. PlantUML forum continuously break and the forum rejected my message with the error: Maximum length is 8000 characters, but the text has only 3800 characters)_ 

As mentioned in https://forum.plantuml.net/11406/loadjson-function (in the PR https://github.com/plantuml/plantuml/pull/755), the authentication to URL endpoints was not in the scope of the implementation for %loadJSON.

In the meantime, I have almost finished my implementation to enable two authentication methods for PlantUML URL accesses. One is BasicAuth, the other is OAuth2 with the grant types "client_credentials" and "password".

It is obvious that storing credentials in DSL scripts (in the PUML files) cannot be a solution. Already in other forum discussions were some ideas that you store the credentials in extra configurations. This is also done by other applications, like the CI tool Jenkins. Administrators can create credentials, and users of the tool are allowed to select them by name to call other services with the proper authentication. The user does not know the access data.

Until now, there was no special requirement to authenticate URLs. Usually people used URLs for "include" and "import" which are publicly available. With the new %loadJSON function, however, there is a desire to also access API functions of services that are almost always protected with at least BasicAuth, but increasingly also with OAuth2.  

The implementation I have currently created allows for SURL in general to have authentication performed. This means that in the future this will also work for includes and imports and all pre-processor commands that use URLs as parameters.  

But there are only certain scenarios where you would use that in practice:  

To generate charts together with data from local intranet or internet:
-  PlantUML as local application (CLI)
- PlantUML as dependency in own projects
- PlantUML as local intranet server

It is probably rather not a function for a PlantUML server that should allow anonymous users to create diagrams (because the users will not have access to server configuration - and should not know the config).  

There were, in my opinion, two ways to provide a URL with an authentication that can be transparently loaded from a local configuration. Either one takes the URL itself as a parameter and searches for a matching configuration entry or to tag the URL itself with a configuration parameter to point exactly to a specific authentication method.  

I chose the second method because it has some advantages. You can see immediately that you want authentication, and you don't need complicated URL matching methods and then searching for a matching configuration. You can use the same credentials for completely different hosts and URL addresses in SSO environments.

Examples: 

- https://jenkinsplantuml@jenkins.example.com/api/json?tree=jobs 
- https://jiraplantuml@jira.example.com/rest/agile/1.0/PRJ-3233  

With the two examples, you now require two configuration files, jenkinsplantuml.credentials and jiraplantuml.credentials. The two files are then used to store the access information. PlantUML only looks in one folder (which is configured as a system property) and reads the data from there. Depending on the configuration, the authentication method is then executed.

The configuration is done in JSON.

Example for BasicAuth:

```
{ 
    "name": "jiraplantuml", 
    "type": "basicauth", 
    "identifier": "alice", 
    "secret": "secret"  
} 

```
Example for OAuth2: 

```
{ 
    "name": "oauth-example", 
    "type": "oauth", 
    "identifier": "demo-backend-client", 
    "secret": "MJlO3binatD9jk1", 
    "properties": { 
        "grantType": "client_credentials", 
        "scope": "read write", 
        "accessTokenUri": "https://login-demo.curity.io/oauth/v2/oauth-token"
    } 
} 
```
 

I would be happy if there are still ideas. I'm currently in the testing phase and everything seems to work as I would expect. That means in the next few days I would put a pull request to it.

Best regards, Aljoscha

**Comment by arnaudroques** [comment]
> (I tried to write a forum message, but I gave up. PlantUML forum continuously break and the forum rejected my message with the error: Maximum length is 8000 characters, but the text has only 3800 characters)

Sorry about that... Let's talk here



> Examples:
> 
>     * https://jenkinsplantuml@jenkins.example.com/api/json?tree=jobs
> 
>     * https://jiraplantuml@jira.example.com/rest/agile/1.0/PRJ-3233
> 
> 
> With the two examples, you now require two configuration files, jenkinsplantuml.credentials and jiraplantuml.credentials. The two files are then used to store the access information. PlantUML only looks in one folder (which is configured as a system property) and reads the data from there. Depending on the configuration, the authentication method is then executed.
> 
> The configuration is done in JSON.

All these sounds logical and user friendly, **that's a great idea**.
I have a question about the two example configuration files `jenkinsplantuml.credentials` and `jiraplantuml.credentials` written in JSON.

Did you create this format or were you inspired from some existing standard ?

> PlantUML only looks in one folder (which is configured as a system property)

What did you choose as system property? What about `plantuml.credentials.path`?

Last question: is this working without adding external dependencies (library or jars)? This might be a concern.

**Comment by arittner** [comment]
Hi!

> All these sounds logical and user friendly, that's a great idea.
> I have a question about the two example configuration files jenkinsplantuml.credentials and jiraplantuml.credentials written in JSON.
> 
> Did you create this format or were you inspired from some existing standard ?
> 

It's a mix from some configs in Spring and unified to this requirement. I don't like to use properties files. I thought about YAML or JSON. For no real reason, my choice fell on JSO. If you think YAML or something different makes more sense, I can change it. Anyway, the YAML parser in PlantUML looks a little bit strange, I was not really happy with it. 

>>    PlantUML only looks in one folder (which is configured as a system property)
>What did you choose as system property? What about plantuml.credentials.path?

I chose "plantuml.security.credentials.path" and %getenv hides everything which starts with "plantuml.security.". Maybe we need more security ENV configs in the future and this should not be published over scripts functions.

>Last question: is this working without adding external dependencies (library or jars)? This might be a concern.

Completely handcrafted :-D - not a single new dependency. 

best regards, Aljoscha.

**Reply by arnaudroques** [reply]
> If you think YAML or something different makes more sense, I can change it.

No, that's fine, it was really an open question. So **your** proposal is going to be our new standard for authentication :-)

> Anyway, the YAML parser in PlantUML looks a little bit strange, I was not really happy with it.

Sure, this [YAML parser](https://github.com/plantuml/plantuml/blob/master/src/net/sourceforge/plantuml/yaml/SimpleYamlParser.java) is definitively ugly... We did not find a nice one (as for [JSON](https://github.com/plantuml/plantuml/blob/master/src/net/sourceforge/plantuml/json/Json.java)) under MIT license, so we write a minimal one that works on really basic examples. We're waiting for PR :-)


> I chose "plantuml.security.credentials.path" and %getenv hides everything which starts with "plantuml.security.". Maybe we need more security ENV configs in the future and this should not be published over scripts functions.

Ok, that makes sense.


> Completely handcrafted :-D - not a single new dependency.

Congratulation! That's a very good news.
Now we are just curious to see your PR :-)
But take your time, there is no need to rush here.

Thanks again
